tests/apollo/test_skvbc_preexecution.py

```diff
@@ -44,24 +46,57 @@ def start_replica_cmd(builddir, replica_id):
 
 class SkvbcPreExecutionTest(unittest.TestCase):
 
+    async def send_single_read(self, skvbc, client):
+        req = skvbc.read_req(skvbc.random_keys(1))
+        await client.read(req)
+
     async def send_single_write_with_pre_execution(self, skvbc, client):
         kv = [(skvbc.keys[0], skvbc.random_value()),
               (skvbc.keys[1], skvbc.random_value())]
         reply = await client.write(skvbc.write_req([], kv, 0), pre_process=True)
         reply = skvbc.parse_reply(reply)
         self.assertTrue(reply.success)
 
-    @unittest.skip("unstable")
+    async def run_concurrent_pre_execution_requests(self, skvbc, clients, num_of_requests, write_weight=.90):
+        sent = 0
+        write_count = 0
+        read_count = 0
+        while sent <= num_of_requests:
+            async with trio.open_nursery() as nursery:
+                for client in clients:
+                    if random.random() <= write_weight:
+                        nursery.start_soon(self.send_single_write_with_pre_execution, skvbc, client)
+                        write_count += 1
+                    else:
+                        nursery.start_soon(self.send_single_read, skvbc, client)
+                        read_count += 1
+            sent += len(clients)
+        return read_count + write_count
+
     @with_trio
     @with_bft_network(start_replica_cmd)
-    async def test_single_pre_process_request(self, bft_network):
+    async def test_sequential_pre_process_requests(self, bft_network):
         """
-        Ensure that we can create a block using pre-execution feature and retrieve its KV pairs through
-        GetBlockData API request.
+        Use a random client to launch one pre-process request in time and ensure that created block are as expected.
         """
         bft_network.start_all_replicas()
         skvbc = kvbc.SimpleKVBCProtocol(bft_network)
 
         for i in range(100):
             client = bft_network.random_client()
             await self.send_single_write_with_pre_execution(skvbc, client)
+
+    @with_trio
+    @with_bft_network(start_replica_cmd)
+    async def test_concurrent_pre_process_requests(self, bft_network):
+        """
+        Launch concurrent requests from different clients in parallel. Ensure that created block are as expected.
+        """
+        bft_network.start_all_replicas()
+        skvbc = kvbc.SimpleKVBCProtocol(bft_network)
+
+        max_concurrency = len(bft_network.clients) // 2
+        clients = bft_network.random_clients(max_concurrency)
+        num_of_requests = 50 * len(clients)
```

Comment:
Why `50 *`? Won't this result in most requests being dropped, because we can only pre-process up to `num_clients` requests?

Reply:
Changed.